base/applications/rapps/asyncinet.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * PROJECT:     ReactOS Applications Manager
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Async Internet operation using WinINet
+ * COPYRIGHT:   Copyright 2020 He Yang            (1160386205@qq.com)
+ */
+
+#include "rapps.h"
+#include <windows.h>
+#include <wininet.h>
+#include <atlbase.h>
+#include "asyncinet.h"
+
+BOOL AsyncInetAcquire(pASYNCINET AsyncInet);
+VOID AsyncInetRelease(pASYNCINET AsyncInet);
+int AsyncInetPerformCallback(pASYNCINET AsyncInet,
+    ASYNC_EVENT Event,
+    WPARAM wParam,
+    LPARAM lParam
+    );
+VOID CALLBACK AsyncInetStatusCallback(
+    HINTERNET hInternet,
+    DWORD_PTR dwContext,
+    DWORD dwInternetStatus,
+    LPVOID lpvStatusInformation,
+    DWORD dwStatusInformationLength
+    );
+VOID AsyncInetReadFileLoop(pASYNCINET AsyncInet);
+BOOL AsyncInetCleanUp(pASYNCINET AsyncInet);
+VOID AsyncInetFree(pASYNCINET AsyncInet);
+
+pASYNCINET AsyncInetDownload(LPCWSTR lpszAgent,
+    DWORD   dwAccessType,
+    LPCWSTR lpszProxy,
+    LPCWSTR lpszProxyBypass,
+    LPCWSTR lpszUrl,
+    BOOL bAllowCache,
+    ASYNCINET_CALLBACK Callback,
+    VOID* Extension
+    ) // allocate memory for AsyncInet and start a download task
+{
+    pASYNCINET AsyncInet = NULL;
+    BOOL bSuccess = FALSE;
+    DWORD InetOpenUrlFlag = 0;
+    INTERNET_STATUS_CALLBACK OldCallbackFunc;
+
+    AsyncInet = (pASYNCINET)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(ASYNCINET));
+    if (!AsyncInet)
+    {
+        OutputDebugStringA("At File: " __FILE__ " HeapAlloc returned 0\n");
+        return 0;
+    }
+
+    AsyncInet->bCleanUp = FALSE;
+    AsyncInet->bCancelled = FALSE;
+
+    AsyncInet->hInternet = NULL;
+    AsyncInet->hInetFile = NULL;
+
+    AsyncInet->Callback = Callback;
+    AsyncInet->Extension = Extension;
+
+    AsyncInet->hEventHandleCreated = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    InitializeCriticalSection(&(AsyncInet->CriticalSection));
+    AsyncInet->ReferenceCnt = 0;
+    AsyncInet->hEventRefZero = CreateEvent(NULL, TRUE, TRUE, NULL);
+    AsyncInet->hEventHandleClose = CreateEvent(NULL, FALSE, FALSE, NULL);
```

Comment:
What about converting `AsyncInet` to a class?
```
class CAsyncInet:
    public CComObjectRootEx<CComMultiThreadModelNoCS>,
    public IUnknown
{
public:
~CAsyncInet()
{
    // cleanup every handle / pointer here
}
// all async functions / properties here.
};
```
Now you can create it like this:
```
CComObject<CAsyncInet>* pAsync;
HRESULT hr = CComObject<CAsyncInet>::CreateInstance(&pAsync);
if (FAILED_UNEQPECTEDLY(hr))
    return hr;
```
(Please note, at this point reference count is 0, so someone needs to AddRef it!)

And everyone that wants to use it, uses it like:
```
pAsync->AddRef();
// do something
pAsync->Release();
```

Or, since we are using c++:
```
void doSomething(pAsync)
{
    CComPtr<CAsyncInet> pRef = pAsync;   // Assigning to the smart pointer will call AddRef()
    // do something with pRef
}  // When the smart pointer is destroyed here, it will call Release()
```

Reply:
The first call to Release where reference count reaches 0 will destroy the object.